File: ios/DesMoinesInsider/Services/SurpriseMeService.swift
import Foundation
import CoreLocation

/// Client for the get_surprise_pick RPC + outcome tracking. Powers the
/// "Surprise Me" button on the iOS Home tab (IOS-DISCOVER-2026-008).
@MainActor
@Observable
final class SurpriseMeService {
    static let shared = SurpriseMeService()

    struct Pick: Identifiable, Decodable, Hashable {
        var id: String { itemId.uuidString + ":" + itemType }
        let itemType: String
        let itemId: UUID
        let title: String?
        let description: String?
        let imageUrl: String?
        let reason: String
        let reasonTemplate: String?

        enum CodingKeys: String, CodingKey {
            case itemType = "item_type"
            case itemId = "item_id"
            case title
            case description
            case imageUrl = "image_url"
            case reason
            case reasonTemplate = "reason_template"
        }
    }

    enum Outcome: String { case shown, saved, tried_another, opened }

    private let supabase = SupabaseService.shared.client

    private init() {}

    func surprise(at location: CLLocation? = nil) async throws -> Pick {
        guard let client = supabase else {
            throw NSError(domain: "SurpriseMe", code: -1, userInfo: [NSLocalizedDescriptionKey: "Not configured"])
        }
        struct Params: Encodable {
            let p_user_lat: Double?
            let p_user_lon: Double?
        }
        let params = Params(
            p_user_lat: location?.coordinate.latitude,
            p_user_lon: location?.coordinate.longitude,
        )
        let rows: [Pick] = try await client
            .rpc("get_surprise_pick", params: params)
            .execute()
            .value
        guard let pick = rows.first else {
            throw NSError(domain: "SurpriseMe", code: 404, userInfo: [NSLocalizedDescriptionKey: "No surprise pick available"])
        }
        // Fire-and-forget shown event for acceptance-rate analytics
        Task { try? await track(pick: pick, outcome: .shown) }
        return pick
    }

    func track(pick: Pick, outcome: Outcome) async throws {
        guard let client = supabase else { return }
        struct Row: Encodable {
            let item_type: String
            let item_id: UUID
            let outcome: String
            let reason_template: String?
        }
        let row = Row(
            item_type: pick.itemType,
            item_id: pick.itemId,
            outcome: outcome.rawValue,
            reason_template: pick.reasonTemplate,
        )
        try await client
            .from("surprise_pick_outcomes")
            .insert(row)
            .execute()
    }
}

File: android/app/src/main/java/com/desmoines/aipulse/data/remote/SurpriseMeService.kt
package com.desmoines.aipulse.data.remote

import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.rpc
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Client for the get_surprise_pick RPC + outcome tracking.
 * Mirrors iOS SurpriseMeService.swift.
 */
@Singleton
class SurpriseMeService @Inject constructor(
    private val supabaseClient: SupabaseClient?,
) {

    @Serializable
    data class Pick(
        @SerialName("item_type") val itemType: String,
        @SerialName("item_id") val itemId: String,
        val title: String? = null,
        val description: String? = null,
        @SerialName("image_url") val imageUrl: String? = null,
        val reason: String,
        @SerialName("reason_template") val reasonTemplate: String? = null,
    )

    enum class Outcome(val value: String) {
        SHOWN("shown"),
        SAVED("saved"),
        TRIED_ANOTHER("tried_another"),
        OPENED("opened"),
    }

    @Serializable
    private data class RpcParams(
        val p_user_lat: Double? = null,
        val p_user_lon: Double? = null,
    )

    @Serializable
    private data class OutcomeRow(
        val item_type: String,
        val item_id: String,
        val outcome: String,
        val reason_template: String? = null,
    )

    suspend fun surprise(latitude: Double? = null, longitude: Double? = null): Pick {
        val client = supabaseClient
            ?: throw IllegalStateException("Supabase not configured")
        val rows = client.postgrest.rpc(
            function = "get_surprise_pick",
            parameters = RpcParams(p_user_lat = latitude, p_user_lon = longitude),
        ).decodeList<Pick>()
        val pick = rows.firstOrNull() ?: throw IllegalStateException("No surprise pick available")
        // Fire-and-forget shown event
        runCatching { track(pick, Outcome.SHOWN) }
        return pick
    }

    suspend fun track(pick: Pick, outcome: Outcome) {
        val client = supabaseClient ?: return
        client.from("surprise_pick_outcomes").insert(
            OutcomeRow(
                item_type = pick.itemType,
                item_id = pick.itemId,
                outcome = outcome.value,
                reason_template = pick.reasonTemplate,
            )
        )
    }
}
